kmip/services/server/engine.py

```diff
@@ -1509,14 +1509,20 @@ def _process_destroy(self, payload):
         else:
             unique_identifier = self._id_placeholder
 
-        self._get_object_with_access_controls(
+        managed_object = self._get_object_with_access_controls(
             unique_identifier,
             enums.Operation.DESTROY
         )
 
-        # TODO (peterhamilton) Process attributes to see if destroy possible
-        # 1. Check object state. If invalid, error out.
-        # 2. Check object deactivation date. If invalid, error out.
+        # TODO If in "ACTIVE" state, we need to check its "Deactivation date"
+        # to see whether it can be destroyed or not
+        if hasattr(managed_object, 'state'):
+            if managed_object.state != enums.State.PRE_ACTIVE and \
+               managed_object.state != enums.State.DEACTIVATED:
+                raise exceptions.PermissionDenied(
+                    "Object is not pre-active or deactivated "
+                    "and cannot be destroyed."
+                )
```

Comment:
Let's also handle the case where if the state is `COMPROMISED` we move it to `DESTROYED COMPROMISED`.